# 5. Redondeo

## 5.1 -i) 

####Número real positivo x con precisión infinita

In [18]:
x=pi

π = 3.1415926535897...

#### truncamos pi a 4 cifras decimales, de modo que     

####△(pi)= 3.1416

####▽(pi)= 3.1415

In [2]:
bits(0.1)

"0011111110111001100110011001100110011001100110011001100110011010"

## 5.1 -ii) 

####Para arítmetica flotante de IEEE:

##### Pasamos x = 0.1 a binario usando el programa decimal_con_punto_a_binario(x), obtenemos 

####decimal_con_punto_a_binario(0.1) = 0.0001100110011001100110011001100110011001100110011001101 

####1. bit del signo: "0"  

####2. 1.1 0011 0011 0011... * 2 ^-4. Calculando el exponente en binario, tenemos  -4 + 127 = 123 = 1111011  

####3. El significante es 10011 0011 0011  

####Por consiguiente, 

####0 (signo)| 1111011 (exponente) |  (significante) 10011 0011 0011 0011 0011 001 = 0 1111011  10011 0011 0011 0011 0011 001

#### donde se truncó a 32 bits de precisión.

####Luego, los redondeos hacia arriba y hacia abajo son 

####△(x) = 0 1111011 10011 0011 0011 0011 0011 010

####▽(x) = 0 1111011 10011 0011 0011 0011 0011 001 

####cuya diferencia es
                                                   0.000000000001
                                                   
####Nota: Si comparamos el redondeo hacia arriba obtenido aqui con el resultado dado por "bits", notamos que estos coinciden al final, por lo que la función bits redondea por defecto hacia arriba.

## 5.2 

####Sean  x_1 = 1.1  y  x_2 = 10.1 , 

In [3]:
bits(1.1)

"0011111111110001100110011001100110011001100110011001100110011010"

##### function decimal_con_punto_a_binario(1.1) = 1.0001100110011001100110011001100110011001100110011 ... 
####                                                                              = 1.000 1100 1100 1100 1100 1100... * 2^0

#### signo: 0
#### exponente: 0 + 127= 127 -> 1111111 
#### significante a 32 bits de precisión: 000 1100 1100 1100 1100 1100 1 

#### Juntando todo obtenemos

#### 0 111111 000 1100 1100 1100 1100 1100 1 

#### redondeamos hacia arriba y hacia abajo                                                                                                                  

#### △(x) = 0 111111 000 1100 1100 1100 1100 11010        

#### ▽(x) = 0 111111 000 1100 1100 1100 1100 11001

#### notamos de nuevo que el redondeo hacia arriba es el que aplica por defecto la compu. 

In [8]:
bits(10.1)

"0100000000100100001100110011001100110011001100110011001100110011"

In [9]:
bits(130)

"0000000000000000000000000000000000000000000000000000000010000010"

##### decimal_con_punto_a_binario(10.1) =  1010.0001100110011001100110011001100110011001100110011 ...
####                                                                 = 1.010 000 1100 1100 1100 1100 1100 1100 1100 1100...*2 ^ 3

#### signo: 0 
#### exponente: 3+127= 130 = 10000010
#### significante a 32 bits de precisión:  010 000 1100 1100 1100 1100 1

#### juntando todo obtenemos

#### 0 100000010 010 000 1100 1100 1100 11001

#### redondeamos hacia arriba y hacia abajo, 

#### △(x) = 0 100000010 010 000 1100 1100 1100 11010          

#### ▽(x) = 0 100000010 010 000 1100 1100 1100 11001

## 5.3 

#### Si x pertenece al subconjunto de los F* entonces x es el mismo número redondeado

## 5.4 

#### En la definición de redondeo hacia arriba, al escoger -x tenemos que                                                                      

#### △(-x) := min { y ∈ F* : y ≥ -x } := max { y ∈ F* : y ≤ x} = ▽(x)  

#### que es la definición de redondeo hacia abajo. Por lo tanto,                                                                                                   

#### △(-x) = ▽(x)

## 5.5

#### Sean los números
####                                                                          x= 1.10 * 2^1 e y = 1.01 * 2 ^ -1
#### De esta manera, 
####                                                                          x + y = 11.0 + 0.101 = 11.101 = 1.1101 * 2 ^1 

#### donde dicho resultado no pertenece al conjunto de números de F.

## 5.6

#### Para que el resultado anterior sea un número que pertenezca al conjunto F, es necesario truncar x+y, que para éste caso se hace escogiendo los dos primeros dígitos después del punto, obteniendo

####                                                                                          x + y = 1.11 *2^1

## 5.7 

####La regla x + y = x + y' se cumple para y  y' no se cumple para los números de punto flotante. Por ejemplo: 

#### Sean los números en punto flotante                                                                 
####                                                                     x =  1.10 * 2^1 = 11.0 , y = 1.01 * 2^-1 = 0.101 .   

#### Por consiguiente,                                                   

####                         x + y = 11.0 + 0.101 = 11.101 = 1.1101 * 2^1 (notación científica) = 1.11 * 2^1 (notación de punto flotante). 

#### En notación de punto flotante solo se admiten dos números binarios después del punto, razón por la cual en el ejemplo anterior se omitieron el par de números "01".   

#### Ahora, sea                                                           y' = 1.10 * 2^-1 = 0.110 

#### entonces,                                              
####                                                   x + y' = 11.0 + 0.110 = 11.110 = 1.1110 * 2^1 = 1.11* 2^1  

#### Por lo tanto, se puede apreciar que x + y = x + y' siendo y diferente de y'

## 5.8

In [17]:
#function f(::)
x_0=1/10;
for i=1:10
    y=mod(3.0^i*x_0,1)
    println(y)
    end 
end

0.30000000000000004
0.8999999999999999
0.7000000000000002
0.09999999999999964
0.3000000000000007
0.9000000000000057
0.700000000000017
0.10000000000002274
0.3000000000001819
0.9000000000005457


LoadError: syntax: unexpected end
while loading In[17], in expression starting on line 7

#### Numéricamente, en principio el cuarto valor  debería ser 0.1=1/10 pero como 0.1 está aproximado, al hacer las iteraciones se van acumulando errores.
#### Analíticamente, volvemos al valor 1/10. 

## 5.9

In [76]:
sum([1/i^2 for i=1:1000]) #guarda los elementos de la serie en un arreglo y luego muestra el resultado de la suma de éstos

1.6439345666816008

#### 1.643 934 566 681 600 (8)

## 5.10

#### Resolviendo las integrales de la expresión dada, obtenemos que

#### 1/(N+1) < Tn < 1/N

In [92]:
N= 1000;
1/(N+1)

0.000999000999000999

In [1]:
N= 1000;
1/N

0.001

## 5.11

In [77]:
set_rounding(Float64, RoundUp) #cota superior con precisión simple
1.6439345666816008

1.643934566681601

#### 1.643 934 566 681 601

In [78]:
set_rounding(Float64, RoundDown) #cota inferior con precisión simple
1.6439345666816008

1.6439345666816008

#### 1.643 934 566 681 600 (8)

## 5.12

#### Juntando toda la información, tenemos que las cotas para Sn y Tn son

####  Sn -> [1.643 934 566 681 600, 1.643 934 566 681 601] 

####  Tn -> [0.0009, 0.001]

#### Ya que S= Sn + Tn, entonces las cotas garantizadas para S son

#### S ->  [1.644 834 566 681 5998, 1.644 934 566 681 6009]

#### Confirmando si el valor real se encuentra dentro de las cotas de S, se tiene

####                                                                                         1.644  834  566  681  5998  
####  [1.644 834 566 681 5998, 1.644 934 566 681 6009] = 1.643  934  566  681  6008
####                                                                                         1.644  934  566  681  6009  
                                                
#### el número está dentro de las cotas

In [88]:
# se suman las cotas inferiores de Sn y Tn para dar la cota inferior garantizada de S
1.643934566681600 + 0.0009

1.6448345666815998

In [89]:
# se suman las cotas superiores de Sn y Tn para dar la cota superior garantizada de S
1.643934566681601+0.001

1.6449345666816009

## 5.13

In [4]:
set_bigfloat_precision(100)

100

In [5]:
BigFloat(sum([1/i^2 for i=1:1000]))

1.6439345666815594615428608449292e+00 with 100 bits of precision

#### 1.643 934 566 681 520 825 781 603 889 4816 e+00

In [55]:
set_rounding(Float64, RoundDown)
1.6439345666815208257816038894816e+00

1.6439345666815206

#### 1.643 934 566 681 520 (6)

In [56]:
set_rounding(Float64, RoundUp)
1.6439345666815208257816038894816e+00

1.6439345666815208

#### 1.643 934 566 681 520 (8)    

In [6]:
N= 1000;
BigFloat(1/(N+1))

9.9900099900099900013250575625534e-04 with 100 bits of precision

In [7]:
N=1000;
BigFloat(1/N)

1.0000000000000000208166817117217e-03 with 100 bits of precision

#### Juntando toda la información, tenemos que las cotas para Sn y Tn son

####  Sn -> [1.643 934 566 681 520 (6),  1.643 934 566 681 520 (8) ] 

####  Tn -> [9.9900099900099900013250575625534e-04, 1.0000000000000000208166817117217e-03]

#### Ya que S= Sn + Tn, entonces las cotas garantizadas para S son

#### S ->  [1.644 834 566 681 5998, 1.644 934 566 681 6009]

#### Confirmando si el valor real se encuentra dentro de las cotas de S, se tiene

####                                                                                             1.644  933  567  680 5217  
####  [1.644 933 567 680 5217, 1.644 934 566 681 5207] =     1.643  934  566  681  55946 
####                                                                                             1.644  934  566  681  5207  
                                                 

In [8]:
1.6439345666815206+9.9900099900099900013250575625534e-04

1.6449335676805217

In [9]:
1.6439345666815208+1.0000000000000000208166817117217e-03

1.6449345666815207

## 5.14

In [15]:
sum([1/(1000-i)^2 for i=1:999]) #se intercambia el orden en el que se suman los valores

1.6439335666815598

In [ ]:
# 1.6439345666816008 comparación con el valor obtenido anteriormente para precisión simple

In [1]:
N= 1000;
1/(N+1)

0.000999000999000999

In [2]:
N=1000;
1/N

0.001

In [3]:
set_rounding(Float64, RoundUp) #cota superior con precisión simple
1.6439335666815598

1.64393356668156

In [4]:
set_rounding(Float64, RoundDown) #cota superior con precisión simple
1.6439335666815598

1.6439335666815598

#### juntando toda la información obtenida,

#### Sn -> [1.6439335666815598, 1.64393356668156]

#### Tn -> [0.000999000999000999, 0.001]

#### Ya que S= Sn + Tn, entonces las cotas garantizadas para S son

####                                                                                        1.643 933 566 681 5598
#### S -> [1.6449325676805606, 1.6449335666815599] = 1.643 933 566 681 5598
####                                                                                        1.644 933 566 681 5599

In [5]:
1.6439335666815598+0.000999000999000999

1.6449325676805606

In [6]:
1.64393356668156+0.001

1.6449335666815599

#### Si cambian las cotas obtenidas anterioremente y la mejor manera de hacerlo es con los límites invertidos ya que así disminuye el error en el intervalo.

## 5.15  

In [7]:
sum([1/i for i=1:1000])

7.485470860550223

In [8]:
set_rounding(Float64, RoundUp) #cota superior con precisión simple
7.485470860550223

7.485470860550223

In [9]:
set_rounding(Float64, RoundDown) #cota superior con precisión simple
7.485470860550223

7.4854708605502225

#### por lo tanto las cotas de la función son

####                                                                             7.485 470  860 550 222
#### [7.4854708605502225, 7.485470860550223] = 7.485 470 860 550 223
####                                                                              7.485 470 860 550 223

## Algunos apuntes de clase

In [3]:
x=0.1

0.1

##### guardan en x el valor ya redondeado (convertido a un flotante cercano), segun el modo actual de redondeo

In [5]:
get_rounding(Float64)

RoundingMode(0)

In [6]:
bits(x)

"0011111110111001100110011001100110011001100110011001100110011010"

##### a partir de ahi sacar el exponente, la mantisa, etc. Para saber a que numero decimal corresponde

In [11]:
set_rounding(Float64, RoundDown)

0

In [9]:
float("0.1")

0.1

In [10]:
big(0.1)

1.000000000000000055511151231257827021181583404541015625e-01 with 256 bits of precision

##### tiene 17 decimales de precision (se cuenta hasta el ultimo cero)

In [19]:
0.1*1.25

0.12499999999999999

In [21]:
set_rounding(Float64, RoundUp)
0.1*1.25

0.12500000000000003

In [22]:
set_rounding(Float64, RoundDown)
0.1*1.25

0.12499999999999999

In [12]:
bits(0.1*1.25)

"0011111110111111111111111111111111111111111111111111111111111111"

In [13]:
set_rounding(Float64, RoundDown)
0.1*1.25

0.12499999999999999

In [14]:
bits(ans)

"0011111110111111111111111111111111111111111111111111111111111111"